src/main/java/frc/robot/subsystems/internals/InternalSubsystem.java

```diff
@@ -92,79 +99,60 @@ public InternalSubsystem(TurretSubsystem turretSubsystem) {
-        //boolean entranceDetected = entrance.get() >= 0.4;
-        //boolean stagingDetected = staging.get() >= 0.2;
-
-        double e = entrance.get();
-        double s = staging.get();
-        boolean entranceDetected = e >= 0.4;
-        boolean stagingDetected = s >= 0.2;
-
-        /*
-        System.out.println("Entrance: " + e);
-        System.out.println("Staging: " + s);
-
-        System.out.println("STORAGE matched " + (
-            storageColor == RED ? "RED" :
-            storageColor == BLUE ? "BLUE" :
-            storageColor == EMPTY ? "EMPTY" :
-            "null"));
-        */
-
-        if (driverOverride) return;
-
-        // Check the entrance sensor for incoming balls and update the ball count accordingly
-        if (!prevEntranceDetected && entranceDetected) ballCount++;
+        // Check sensors for incoming and exiting balls and update ball counts accordingly
+        boolean entranceDetected = entrance.get() >= 0.4;
+        if (!prevEntranceDetected && entranceDetected) entranceStorageBallCount++;
         prevEntranceDetected = entranceDetected;
 
-        // If a ball has entered internals, start the bottom motor
-        if (entranceDetected) motorBottom.set(0.3);
-
-        // If there is a ball in storage, start the top motor
-        if (isBall(storageColor)) {
+        Color storageColor = matchColor(colorSensorThread.getLastStorage());
+        if (!prevStorageDetected && isBall(storageColor)) {
             // If we haven't already checked rejection logic, reject the ball if it doesn't match alliance color
             if (!rejectingChecked) {
                 rejecting = storageColor != ALLIANCE_COLOR;
                 rejectingChecked = true;
             }
-            motorTop.set(0.5);
-            //System.out.println("STORAGE detected, running top motor");
+            entranceStorageBallCount--;
+            storageStagingBallCount++;
         }
 
-        // If there is a ball in staging, stop the bottom and top motors
-        if (stagingDetected) {
-            motorBottom.set(0);
-            motorTop.set(0);
-            //System.out.println("STAGING detected, stopping both motors");
+        boolean stagingDetected = staging.get() >= 0.2;
+        if (!prevStagingDetected && stagingDetected) {
+            storageStagingBallCount--;
+            stagingExitBallCount++;
         }
+        prevStagingDetected = stagingDetected;
+
+        if (driverOverride) return;
+
+        // If there is a ball in the entrance or between storage and staging *and* staging is empty, run the bottom motor.
+        // This serves to bring the first ball properly from storage to staging. For the second ball, the right hand side
+        // of the condition will evaluate to false so it will be stopped when the ball reaches storage.
+        motorBottom.set(entranceStorageBallCount > 0 || (stagingExitBallCount == 0 && storageStagingBallCount > 0)
```

Comment:
what if there's a ball already in storage?

Reply:
Is it possible to have a ball in storage without intake running?
(like at the start of the match)

Reply:
yes, we will have one at the start of the match. we can code for that and just assume at the beginning/initialization our count is 1.
and for the motors, we have tested that before, if we redeploy code with a ball in storage or staging (where the motors will be off anyways) it doesn't mess anything up.

Reply:
If we could make the logic resilient enough so that when we run the intake in autonomous to pick up a second (and therefore run the bottom roller), the ball in the robot hits storage before the second passes thru the entrance sensor and the software can recognize that